### PR TITLE
Review and refactor persistent list implementation

### PR DESCRIPTION
**Pull request by ilya-g**

### CODE REVIEW

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt [hunk 1]
```diff
@@ -17,253 +17,289 @@
...
         }
-        val lastBufferWrapper = ObjectWrapper(null)
-        val newRoot = pullLastBuffer(root, shift, rootSize - 1, lastBufferWrapper)!!
-        val newTail = lastBufferWrapper.value as Array<Any?>
+        val tailCarry = ObjectWrapper(null)
+        val newRoot = pullLastBuffer(root, shift, rootSize - 1, tailCarry)!!
+        @Suppress("UNCHECKED_CAST")
+        val newTail = tailCarry.value as Array<Any?>
 
+        // check if the new root contains
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt]:
```suggestion
        // check if the new root contains only one element
```

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt [hunk 2]
```diff
@@ -117,120 +139,131 @@ internal class PersistentVector<E>(private val root: Array<Any?>,
...
-            tailWrapper.value = root[0]
+            root.copyInto(newRoot, bufferIndex, bufferIndex + 1, MAX_BUFFER_SIZE)
+            newRoot[MAX_BUFFER_SIZE - 1] = tailCarry.value
+            tailCarry.value = root[0]
             return newRoot
         }
 
+        // Could be val bufferLastIndex = root.indexOfLast { it != null }, but that isn't optimized enough
         var bufferLastIndex = MAX_BUFFER_SIZE_MINUS_ONE
         while (root[bufferLastIndex] == null) {
```
Comment (qurbonzoda) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt]: can be optimised to:
```
var bufferLastIndex = indexSegment(rootSize() - 1, shift)
if (bufferLastIndex < MAX_BUFFER_SIZE - 1 && root[bufferLastIndex + 1] != null) {
    bufferLastIndex = MAX_BUFFER_SIZE - 1
}
```

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt [hunk 3]
```diff
@@ -220,131 +229,142 @@ class PersistentVectorBuilder<E>(private var vector: PersistentList<E>,
...
-            return mutableRest
+            val mutableRoot = makeMutable(root)
+            root.copyInto(mutableRoot, bufferIndex, bufferIndex + 1, MAX_BUFFER_SIZE)
+            mutableRoot[MAX_BUFFER_SIZE - 1] = tailCarry.value
+            tailCarry.value = removedElement
+            return mutableRoot
         }
 
         var bufferLastIndex = MAX_BUFFER_SIZE_MINUS_ONE
         while (root[bufferLastIndex] == null) {
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt]: ok, please apply these two loop-removal optimizations in a separate commit after this PR.

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt (hunk 2 above)
Comment (ilya-g) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt]: ok, please apply these two loop-removal optimizations in a separate commit after this PR.

**Pull request merged**